### PR TITLE
Prevalidate node removal API (pt. 1)

### PR DESCRIPTION
**Pull request by pxsalehi**
This PR adds the first part of the Prevalidate Node Removal API. This
API allows checking  whether attempting to remove some node(s) from the
cluster is likely to succeed or not. This check is useful when a node
needs to be removed from a RED cluster, without risking loosing the last
copy of some RED shards.

In this PR, we only check whether a RED index is a Searchable Snapshot
index or not, in which case the removal of any node is safe as the RED
index is backed by a snapshot.

Relates https://github.com/elastic/elasticsearch/issues/87776

### CODE REVIEW

**Comment by pxsalehi** [comment]
@elasticmachine run elasticsearch-ci/docs

**Comment by pxsalehi** [comment]
API doc: https://elasticsearch_88952.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/prevalidate-node-removal-api.html

**Comment by pxsalehi** [comment]
@henningandersen Please have a look at this draft PR. I have tried to find out all the places where the new API would "touch", and add a basic version of what is needed. As far as the prevalidation logic goes, it is for now as basic as: if red cluster then it might not be safe to remove any node, and otherwise any node is safe to remove (essentially what is currently implied by the cluster health status).

Please check if the naming/location of the API related files/doc, the doc itself, and whether there is anything missing from the structure/scaffolding of adding a new API. I'll try to add more logic/validation/test to the actual API implementation, and refine the doc.

File: server/src/internalClusterTest/java/org/elasticsearch/cluster/PrevalidateNodeRemovalIT.java
Comment (idegtiarenko) [inline server/src/internalClusterTest/java/org/elasticsearch/cluster/PrevalidateNodeRemovalIT.java:32]: NIT: is it possible to replace it with a single assertion? This way in case of failure we would be able to see the entire map content without having to reproduce it.

```
assertThat(resp.getPrevalidation().getPerNodeResult(), equalTo(Map.of(nodeName, NodesRemovalPrevalidation.IsSafe.YES)));
```

I was hoping there is a hamcrest matches for it, but it looks like `hasEntry` permits other entries in map as well.

Comment (idegtiarenko) [reply server/src/internalClusterTest/java/org/elasticsearch/cluster/PrevalidateNodeRemovalIT.java:32]: Oh, looks like the entry will contain a reason as well. Was it not asserted intentionally?

Comment (pxsalehi) [reply server/src/internalClusterTest/java/org/elasticsearch/cluster/PrevalidateNodeRemovalIT.java:32]: Yes. For now those details are not clear. I just added that IT to make sure everything works together. There will be more tests and detailed assertions, ofc.

File: docs/reference/cluster/prevalidate-node-removal.asciidoc
Comment (henningandersen) [inline docs/reference/cluster/prevalidate-node-removal.asciidoc:7]: Not sure if this is meant to be removed before merge. We normally do the `{cloud-only}` warning only.

File: docs/reference/cluster/prevalidate-node-removal.asciidoc
Comment (henningandersen) [inline docs/reference/cluster/prevalidate-node-removal.asciidoc:36]: I think we can support node-ids too, but we should support node name and node external-id (which is sort of the same).

This avoids orchestration systems having to lookup the node-id first. The lookup of node-id would need to also by-pass circuit breakers. And limiting the number of calls also increases the chance of success.

With two identification forms it might be easier to switch to a body for the input?

File: docs/reference/cluster/prevalidate-node-removal.asciidoc
Comment (henningandersen) [inline docs/reference/cluster/prevalidate-node-removal.asciidoc:45]: The naming here looks different from existing naming. Maybe you can look for a few examples to lean on?

I'd tentatively think that the outer level result needs no `overall` name, but could still be an object `result`.

`per_node_result` could be just `nodes` with a list of nodes containing (id, name, result) for each.

File: docs/reference/cluster/prevalidate-node-removal.asciidoc
Comment (henningandersen) [inline docs/reference/cluster/prevalidate-node-removal.asciidoc:50]:
```suggestion
This example validates whether it is safe to remove the nodes `node1` and `node2`. The response indicates that it is safe to remove `node1`, but it might not be safe to remove `node2`. Therefore, the overall prevalidation of the removal of the two nodes is `UNKNOWN`.
```

File: rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json
Comment (henningandersen) [inline rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json:4]: We should fill in the URL here.

File: rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json
Comment (henningandersen) [inline rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json:17]: I think POST is slightly better here, since it mostly runs a function (prevalidate), rather than GETs state. We could support both at least.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/remove/PrevalidateNodeRemovalAction.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/remove/PrevalidateNodeRemovalAction.java:16]: I am tempted to use:

```suggestion
    public static final String NAME = "cluster:admin/shutdown/prevalidate_removal";
```

This seems more shutdown related than "nodes" related.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/remove/PrevalidateNodeRemovalResponse.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/remove/PrevalidateNodeRemovalResponse.java:23]: Maybe this is just because it is draft? I see no reason this response could not be sent over the wire.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/remove/TransportPrevalidateNodeRemovalAction.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/remove/TransportPrevalidateNodeRemovalAction.java:30]: Yes.

File: rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json
Comment (henningandersen) [inline rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json:19]: Since it is a master action, there are some timeouts to add here.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/remove/TransportPrevalidateNodeRemovalAction.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/remove/TransportPrevalidateNodeRemovalAction.java:46]: `new ClusterStateHealth(clusterService.state()).getStatus()` is preferable here, avoids the async, avoids another action needing to have circuit breaking disabled and ensures we can use the same `ClusterState` for a full prevalidation as we add more features.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/remove/TransportPrevalidateNodeRemovalAction.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/remove/TransportPrevalidateNodeRemovalAction.java:39]: We should utilize `canTripCircuitBreaker` set to false here (easiest to do after you switch to `TransportMasterNodeReadAction`.)

File: x-pack/plugin/security/qa/operator-privileges-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/operator/Constants.java
Comment (henningandersen) [inline x-pack/plugin/security/qa/operator-privileges-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/operator/Constants.java:46]: This should be an operator only action instead, i.e., not added here.

**Review by henningandersen — COMMENTED** [review]
Did a quick walk through of the code and added initial comments.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/TransportPrevalidateNodeRemovalAction.java
Comment (pxsalehi) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/TransportPrevalidateNodeRemovalAction.java:106]: I have tried to add some validations here to make sure prevalidation calls do not accidentally lead to a false safe removal response. By that I mean, since we do provide an overall result, if the provided nodes (name or id) do not each uniquely identify a node, I do not consider the request valid. I am assuming it is not safe to dismiss an id (perhaps with a typo) if there is no matching node for it. I am not sure whether this is strictly necessary, so I'd appreciate a second opinion.

**Review by pxsalehi — COMMENTED** [review]
@henningandersen I've addressed your comments. Please have another quick look. I will cleanup a bit more and add more tests, but for now I have some questions about the API.

About the scope of the PR, do we need to address the second phase (reaching out to the nodes to see if they have a copy of an unassigned shard) in this PR? If possible I'd rather do that as a follow up, as just addressing the searchable snapshot indices for now seems to be also useful.

File: docs/reference/cluster/prevalidate-node-removal.asciidoc
Comment (henningandersen) [reply docs/reference/cluster/prevalidate-node-removal.asciidoc:57]: I'd be happy with just POST.

File: docs/reference/cluster/prevalidate-node-removal.asciidoc
Comment (henningandersen) [reply docs/reference/cluster/prevalidate-node-removal.asciidoc:34]: I did not look at the code yet, but it would be nice if we know what field we expect here to disambiguate.

I think we should not support IP etc., those 3 forms you mention are all we need.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/TransportPrevalidateNodeRemovalAction.java
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/TransportPrevalidateNodeRemovalAction.java]: I think the ambiguous case (more than one node matches) is fine to throw on. I think I'd like to provide a response also if we are missing a node - but the response should indicate that it was missing. That way we can leave it up to the orchestration to make a decision - that might depend on the situation (vacate/system vs user initiated change).

**Comment by henningandersen** [comment]
> About the scope of the PR, do we need to address the second phase (reaching out to the nodes to see if they have a copy of an unassigned shard) in this PR? If possible I'd rather do that as a follow up, as just addressing the searchable snapshot indices for now seems to be also useful.

I agree that we can omit that in this PR and do that as a follow-up as long as it does not result in false positives (i.e., us saying it is ok to shutdown the node when it is really not).

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/PrevalidateNodeRemovalRequest.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/PrevalidateNodeRemovalRequest.java:22]: I'd like to see this split into ids, names and externalIds. Or a flag for the one type we expect (and reject if multiple types are used).

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/PrevalidateNodeRemovalRequest.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/PrevalidateNodeRemovalRequest.java:35]: I wonder why we set this into the request, can we not keep that in a local state instead? I prefer the request to be purely the request and also that writeTo and reading results in the same object.

File: rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json
Comment (henningandersen) [inline rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json:7]: I think we should call this stable before release.

File: server/src/test/java/org/elasticsearch/action/admin/cluster/node/shutdown/PrevalidateNodeRemovalResponseTests.java
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/action/admin/cluster/node/shutdown/PrevalidateNodeRemovalResponseTests.java:78]: Can we use `AbstractWireSerializingTestCase`?

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/PrevalidateNodeRemovalWithSearchableSnapshotTests.java
Comment (henningandersen) [inline x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/PrevalidateNodeRemovalWithSearchableSnapshotTests.java:28]: Conventionally, these should be named either `IT` or `IntegTests`.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/PrevalidateNodeRemovalRequest.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/PrevalidateNodeRemovalRequest.java:29]: Can we add a `AbstractWireSerializingTestCase` for this class too?

File: rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json
Comment (pxsalehi) [reply rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json:7]: Why? What would be the requirement to call it stable? I thought we start with experimental and at some point call it stable. Do we have some guidelines about this?

File: docs/reference/cluster/prevalidate-node-removal.asciidoc
Comment (pxsalehi) [reply docs/reference/cluster/prevalidate-node-removal.asciidoc:57]: If we are moving away from having the nodes in the path (I started with that since I thought it would be similar/consistent to some other APIs we have), should we have three query params and allow setting only one of them per request?

```
POST /_internal/prevalidate_node_removal?name=node1,node2
POST /_internal/prevalidate_node_removal?id=id1,id2
POST /_internal/prevalidate_node_removal?external_id=id1,id2
``` 

This avoids any kind of ambiguity and we can just resolve based on id, name or external ID.

I'm not entirely sure how problematic the length of the URL can be. I am assuming up to 2000 is ok and considering ID/names of length 40, and the fact that the number of id/names is in the order of a fraction of the cluster size, I think not using a body should be ok. WDYT?

Comment (henningandersen) [reply docs/reference/cluster/prevalidate-node-removal.asciidoc:57]: Sounds fine to me.

File: rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json
Comment (henningandersen) [reply rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json:7]: Most features start as stable. In this case, the primary consumer is orchestration and as such we need to be confident enough in the feature to release it GA, i.e., fully support it and only make bwc changes to it in the future.

We can also hide this behind a feature flag until it is ready but I see little value in getting it out as experimental since we will be very unlikely to receive feedback affecting the design.

**Review by henningandersen — COMMENTED** [review]
I did not see any code changes so I simply responded to questions in this review cycle.

**Comment by pxsalehi** [comment]
> I did not see any code changes so I simply responded to questions in this review cycle.

@henningandersen Thanks for the answers. I haven't pushed any code, and the review request was by mistake I guess. :-)

File: docs/reference/cluster/prevalidate-node-removal.asciidoc
Comment (pxsalehi) [reply docs/reference/cluster/prevalidate-node-removal.asciidoc]: I've made some changes. Please let me know if this is still not properly describing the API or not.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/TransportPrevalidateNodeRemovalAction.java
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/TransportPrevalidateNodeRemovalAction.java]: I was going for later finding out the true NOs and separate them from the UNKNOWN. But I think it works both ways. I've changed to NO. So for now we'll have no UKNOWN answer.

File: rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json
Comment (pxsalehi) [reply rest-api-spec/src/main/resources/rest-api-spec/api/_internal.prevalidate_node_removal.json]: Thanks for the explanation. I've changed to stable. Frankly, I imagined we'd leave this experimental until at lease the second part of this API is merged, especially since this is only for internal uses I thought that would be possible w/o hindering adoption. Nonetheless, for now I cannot think of any non-BWC changes that the second part could introduce. It probably adds a timeout parameter and some of those NOs get refined into YES/NO/UNKNOWNs.

File: server/src/internalClusterTest/java/org/elasticsearch/cluster/PrevalidateNodeRemovalIT.java
Comment (pxsalehi) [inline server/src/internalClusterTest/java/org/elasticsearch/cluster/PrevalidateNodeRemovalIT.java:27]: While revisiting this test, I thought of two corner cases which are not tested here, and I am not sure they make sense or not. So I'd like to ask for an opinion.

1. If the cluster is green/yellow, and I send a prevalidation request for all nodes in the cluster, the answer would be YES for the overall result and each node's result! Clearly, we do not mean one could remove all those nodes at once w/o causing any issue, even though such a test would pass! Should we explicitly mention this in the doc, or somehow in case of a non-RED cluster just return no responses. From what I understood this API is for clusters which are RED and could be preceded by a call to the cluster health API. I am more inclined towards clarifying what an overall yes means in the doc (specifically pointing out that it doesn't mean that all nodes could be removed at once).
2. Would it be necessary to validate that all the nodes that are specified are data nodes? Currently, we kind of assume that.

**Review by pxsalehi — COMMENTED** [review]
@henningandersen I've addresses all the comments and have added more tests. I do not have any more changes to add. Please have a look.

File: server/src/internalClusterTest/java/org/elasticsearch/cluster/PrevalidateNodeRemovalIT.java
Comment (henningandersen) [reply server/src/internalClusterTest/java/org/elasticsearch/cluster/PrevalidateNodeRemovalIT.java:27]: Thanks for those explicit questions:

1. Yes, we should clarify that in docs. It is not a replacement for using the shutdown API, it is a pre-validation, answering whether doing so would/should succeed. If there are cases here that are not handled by the shutdown API (might be), we should probably also bake them in (as a follow-up).
2. pre-validation should say yes to any non-data node regardless of shard status. We should still encourage to use the pre-validation API in case we have future requirements for other types of nodes.

Comment (pxsalehi) [reply server/src/internalClusterTest/java/org/elasticsearch/cluster/PrevalidateNodeRemovalIT.java:27]: @henningandersen thanks for the answer. I've added some notes to the docs. The PR is ready for review. 

https://elasticsearch_88952.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/prevalidate-node-removal-api.html

**Comment by pxsalehi** [comment]
While this is waiting for review. I'm gonna start with the second part.

**Comment by pxsalehi** [comment]
As discussed with @henningandersen, I'll remove `UNKNOWN` from the possible answers that the API can return. Therefore, the `is_safe` node result will be a boolean. However, we'll add another field that helps refine the answer in a parsable way, and contains a certain "category" for the yes/no response. I propose renaming the current `reason` field to `message` and using `reason` as the name for the new field. As an example:

```json
result : {
  is_safe: "UNKNOWN",
  reason: "failed contacting the node with ID XXXX: exception ...."
}
```
will change to:
```json
result : {
  is_safe: "NO",
  reason: "no_response_from_node",
  message: "failed contacting the node with ID XXXX: exception ...."
}
```

**Comment by pxsalehi** [comment]
@henningandersen I'm trying to make the change mentioned in my last comment. Since this needs to be changed in several places (parser, tests, etc), I want to make sure I'll get it right this time. I've decided to remove the 'result' top-level object since it is not directly useful there. So the API response would change from this:
```
{
  "result": {
    "is_safe": "UNKNOWN",  // --> or YES, or NO
    "reason": "cluster health is RED"  // --> Just a string
  },
  "nodes": [
    {
      "id": "node1-id",
      "name" : "node1",
      "external_id" : "node1-externalId",
      "result" : {  // -> Same type as the top-level result
        "is_safe": "YES",
        "reason": ""
      }
    }
  ]
}
```
to this:
```
{
  "is_safe": "false",  //--> boolean
  "message": "cluster health is RED"  // --> Just a string
  "nodes": [
    {
      "id": "node1-id",
      "name" : "node1",
      "external_id" : "node1-externalId",
      "result" : {  
        "is_safe": "true",  //--> boolean
        "reason": "no_red_shards",  //--> (just an example) an enum of fixed Strings (useful for parsing/automation)
        "message": "node does not host any red shards" //--> A String that could be anything with details specific to the call/node
      }
    }
  ]
}
```
What do you think about this?

**Comment by henningandersen** [comment]
Thanks @pxsalehi that mostly makes sense, 1-2 comments:

>  ` "is_safe": "false",  //--> boolean`

should be 

>   "is_safe": false,  //--> boolean

Also, renaming reason to message makes sense here, but we could defer the introdution of the reason field to pt.2 (or later) to avoid too many changes in this PR (that is mostly reviewed).

**Comment by pxsalehi** [comment]
@henningandersen Thanks for the feedback. I've made the changes. This is ready for review again. Please have a look.

**Comment by pxsalehi** [comment]
@elasticmachine update branch

**Review by henningandersen — APPROVED** [review]
LGTM

Left a few comments but no need for another review round unless significant changes are made.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/PrevalidateNodeRemovalRequest.java
Comment (pxsalehi) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/shutdown/PrevalidateNodeRemovalRequest.java:52]: Yes. That's not needed anymore. I've removed them.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.prevalidate_node_removal/10_basic.yml
Comment (pxsalehi) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.prevalidate_node_removal/10_basic.yml:31]: These are fields from the `ValidationException` and already exist with those names.

Comment (henningandersen) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.prevalidate_node_removal/10_basic.yml]: Thanks and sorry for the confusion. I simply searched for reason and stumbled upon it, but clearly you are right.

**Comment by pxsalehi** [comment]
@elasticmachine update branch

**Comment by pxsalehi** [comment]
Thanks for the reviews Henning!

**Comment by pxsalehi** [comment]
@elasticmachine run elasticsearch-ci/part-2
@elasticmachine run elasticsearch-ci/part-1
@elasticmachine run elasticsearch-ci/packaging-tests-windows-sample

**Pull request merged**
